### PR TITLE
`xef-reasoning` A new module to encapsulate common patterns with text, code and more.

### PR DESCRIPTION
**Pull request by raulraja**
This pull request introduces the `Xef Reasoning Framework`, which helps encapsulate common interactions with the LLMs. The framework currently focuses on code and text-related capabilities, which are the main use cases we find in clients and demo-related work. It also introduces the ability to make automatic tool selections for a prompt in code interpreter style.

The tool selection comes up with a plan on how to best use a set of tools to solve a problem and runs those tools over the user input attempting to accomplish what the user is asking for.

```kotlin
package com.xebia.functional.xef.auto.reasoning

import com.xebia.functional.xef.auto.CoreAIScope
import com.xebia.functional.xef.auto.llm.openai.OpenAI
import com.xebia.functional.xef.auto.llm.openai.OpenAIEmbeddings
import com.xebia.functional.xef.reasoning.filesystem.Files
import com.xebia.functional.xef.reasoning.pdf.PDF
import com.xebia.functional.xef.reasoning.text.Text
import com.xebia.functional.xef.reasoning.tools.ToolSelection

suspend fun main() {
  val scope = CoreAIScope(OpenAIEmbeddings(OpenAI.DEFAULT_EMBEDDING))
  val model = OpenAI.DEFAULT_CHAT
  val serialization = OpenAI.DEFAULT_SERIALIZATION
  val text = Text(model = model, scope = scope)
  val files = Files(model = serialization, scope = scope)
  val pdf = PDF(chat = model, model = serialization, scope = scope)

  val toolSelection = ToolSelection(
    model = serialization,
    scope = scope,
    tools = listOf(
      text.summarize,
      pdf.readPDFFromUrl,
      files.readFile,
      files.writeToTextFile,
    ),
  )

  val result = toolSelection.applyInferredTools("""|
    |Extract information from https://arxiv.org/pdf/2305.10601.pdf
  """.trimMargin())
  println(result)
}
```
```txt
I 22:46:15.461 [.reasoning.tools.ToolSelection] 🔍 Applying inferred tools for task: Extract information from https://arxiv.org/pdf/2305.10601.pdf
I 22:46:15.461 [.reasoning.tools.ToolSelection] 🔍 Creating execution plan for task: Extract information from https://arxiv.org/pdf/2305.10601.pdf
I 22:46:24.569 [.reasoning.tools.ToolSelection] 🔍 Applying execution plan with reasoning: This execution plan ensures that the PDF is read, the information is summarized, and the summary is written to a text file
I 22:46:24.569 [.reasoning.tools.ToolSelection] 🔍 Applying tool: Read PDF from URL for step: The task requires extracting information from a PDF file located at a specific URL
I 22:46:26.346 [f.reasoning.pdf.ReadPDFFromUrl] Reading url https://arxiv.org/pdf/2305.10601.pdf
I 22:46:28.475 [oning.text.summarize.Summarize] 📚 Summarizing large text of tokens 55971 to approximately Bound(length=1000) tokens
I 22:46:28.678 [oning.text.summarize.Summarize] 📚 Split text into 5 chunks
I 22:46:28.681 [oning.text.summarize.Summarize] 📝 Summarizing chunk with prompt tokens 5145 for length Bound(length=200)
I 22:46:28.681 [oning.text.summarize.Summarize] 📝 Summarizing chunk with prompt tokens 14385 for length Bound(length=200)
I 22:46:28.681 [oning.text.summarize.Summarize] 📝 Summarizing chunk with prompt tokens 7644 for length Bound(length=200)
I 22:46:28.681 [oning.text.summarize.Summarize] 📝 Summarizing chunk with prompt tokens 12421 for length Bound(length=200)
I 22:46:28.682 [oning.text.summarize.Summarize] 📝 Summarizing chunk with prompt tokens 14388 for length Bound(length=200)
I 22:46:29.992 [oning.text.summarize.Summarize] 📝 Summarized chunk in tokens: 34
I 22:46:29.992 [oning.text.summarize.Summarize] 📝 Summarized chunk in tokens: 26
I 22:46:30.034 [oning.text.summarize.Summarize] 📝 Summarized chunk in tokens: 18
I 22:46:31.229 [oning.text.summarize.Summarize] 📝 Summarized chunk in tokens: 114
I 22:46:32.756 [oning.text.summarize.Summarize] 📝 Summarized chunk in tokens: 107
I 22:46:32.757 [oning.text.summarize.Summarize] 📚 Summarized 5 chunks
I 22:46:32.757 [.reasoning.tools.ToolSelection] 🔍 Applying tool: Summarize for step: After reading the PDF, we need to summarize the extracted information
I 22:46:32.757 [oning.text.summarize.Summarize] 📚 Summarizing large text of tokens 324 to approximately Bound(length=1000) tokens
I 22:46:32.758 [oning.text.summarize.Summarize] 📚 Split text into 1 chunks
I 22:46:32.758 [oning.text.summarize.Summarize] 📝 Summarizing chunk with prompt tokens 327 for length Bound(length=1000)
I 22:46:35.507 [oning.text.summarize.Summarize] 📝 Summarized chunk in tokens: 101
I 22:46:35.507 [oning.text.summarize.Summarize] 📚 Summarized 1 chunks
I 22:46:35.507 [.reasoning.tools.ToolSelection] 🔍 Applying tool: Write output to Text file for step: Finally, we need to write the summarized information to a text file

├── Read PDF from URL
│   ├── The task requires extracting information from a PDF file located at a specific URL
│   └── The "Tree of Thoughts" (ToT) framework is introduced to enhance language models' problem-solving abilities. It allows language models to consider multiple reasoning paths and make deliberate decisions by exploring coherent units of text ("thoughts"). ToT significantly improves performance on tasks requiring planning or search. Existing approaches like Chain-of-Thought (CoT) prompting are limited in their exploration and decision-making capabilities. ToT addresses these limitations by decomposing the problem-solving process into thought steps, generating potential thoughts, evaluating states, and using search algorithms. The given text is too long to be summarized within a maximum bound of 200 words. I'm sorry, but the provided text does not seem to be readable or meaningful. It appears to be a string of random characters. Therefore, it cannot be summarized. 1 - The text provided is not readable and does not contain any meaningful information. Therefore, it cannot be summarized.

2 -  1 - The Tree of Thoughts (ToT) framework enhances the problem-solving capabilities of language models (LMs) by empowering them to make autonomous and intelligent decisions. However, future applications involving interaction with external environments or humans could pose potential dangers. On the other hand, ToT improves the interpretability of model decisions and the opportunity for human alignment. LMs address the weakness of classical methods and provide a way to solve complex problems that are not easily formalized. The intersection of LMs with classical approaches to AI is an exciting direction for future work.

2 - 

├── Summarize
│   ├── After reading the PDF, we need to summarize the extracted information
│   └── 1 - The "Tree of Thoughts" (ToT) framework enhances language models' problem-solving abilities by allowing them to consider multiple reasoning paths and make deliberate decisions. ToT improves performance on tasks requiring planning or search by decomposing the problem-solving process into thought steps, generating potential thoughts, evaluating states, and using search algorithms. Existing approaches like Chain-of-Thought (CoT) prompting have limitations in exploration and decision-making capabilities. ToT addresses these limitations and provides a more comprehensive problem-solving approach.

├── Write output to Text file
│   ├── Finally, we need to write the summarized information to a text file
│   └── /var/folders/h7/vh204ss54dv8ghllkvrkhlxr0000gn/T/3e7c7b98-840d-405e-9236-251b4f553b17/summary.txt
```

### CODE REVIEW

**Comment by raulraja** [comment]
@xebia-functional/team-ai

File: core/src/commonMain/kotlin/com/xebia/functional/xef/llm/ChatWithFunctions.kt [hunk 1]
```diff
@@ -8,22 +8,71 @@ import com.xebia.functional.xef.auto.PromptConfiguration
...
 
+  @OptIn(ExperimentalSerializationApi::class)
+  fun generateCFunction(descriptor: SerialDescriptor): List<CFunction> {
+    val fnName = descriptor.serialName.substringAfterLast(".")
+    return generateCFunction(fnName, encodeJsonSchema(descriptor))
+  }
+
+  fun generateCFunction(fnName: String, schema: String): List<CFunction> {
+    return listOf(CFunction(fnName, "Generated function for $fnName", schema))
+  }
```
Comment (diesalbla) [inline core/src/commonMain/kotlin/com/xebia/functional/xef/llm/ChatWithFunctions.kt:35]:
```suggestion
  fun generateCFunction(fnName: String, schema: String): List<CFunction> =
    listOf(CFunction(fnName, "Generated function for $fnName", schema))
```

File: core/src/commonMain/kotlin/com/xebia/functional/xef/llm/ChatWithFunctions.kt [hunk 2]
```diff
@@ -8,22 +8,71 @@ import com.xebia.functional.xef.auto.PromptConfiguration
...
+    context: VectorStore,
+    serializerName: String,
+    jsonSchema: String,
+    conversationId: ConversationId? = null,
+    serializer: (json: String) -> A,
+    functions: List<CFunction> = generateCFunction(serializerName, jsonSchema),
+    promptConfiguration: PromptConfiguration = PromptConfiguration.DEFAULTS,
+  ): A {
+    return prompt(prompt, context, conversationId, functions, serializer, promptConfiguration)
+  }
```
Comment (diesalbla) [inline core/src/commonMain/kotlin/com/xebia/functional/xef/llm/ChatWithFunctions.kt:49]:
```suggestion
  ): A =
    prompt(prompt, context, conversationId, functions, serializer, promptConfiguration)
```

File: reasoning/src/jvmMain/kotlin/com/xebia/functional/xef/reasoning/search/Search.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+) : Tool {
+  override val name: String = "Search"
+
+  override val description: String = "Search the web for the best answer and summarize what's found"
+
+  override suspend fun invoke(input: String): String {
+    val keywords = keywordExtraction.invoke(input)
+    val docs = search(keywords)
+    val combinedDocs = docs.joinToString("\n")
+    return Summarize(model, scope, summaryLength, instructions).invoke(combinedDocs)
```
Comment (diesalbla) [inline reasoning/src/jvmMain/kotlin/com/xebia/functional/xef/reasoning/search/Search.kt:29]: Can this `Summarize` be extracted as a field?

**Review by diesalbla — APPROVED** [review]
couple small comments, but not much to add

File: reasoning/src/jvmMain/kotlin/com/xebia/functional/xef/reasoning/search/Search.kt (hunk 3 above)
Comment (raulraja) [reply reasoning/src/jvmMain/kotlin/com/xebia/functional/xef/reasoning/search/Search.kt:29]: yes, but I'm not accessing it anywhere else but in this method body.

**Pull request merged**